src/main/java/org/primeframework/mvc/util/QueryStringBuilder.java

```diff
@@ -49,51 +49,54 @@ public static QueryStringBuilder builder(String uri) {
   }
 
   public QueryStringBuilder beginFragment() {
-    sb.append("#");
-    addSeparator = false;
+    // If query string contains no terms, remove the question mark
+    if (sb.toString().endsWith("?")) {
+      sb.setLength(sb.length() - 1);
+    }
+
+    if (sb.indexOf("#") == -1) {
+      sb.append("#");
+      addSeparator = false;
```

Comment:
We should just double check that the URL spec never allows more than one `#`. Seems likely but I don't know for sure. 

From my Googling it does appear to be a safe assumption. 

https://www.webdevsplanet.com/post/url-parameters-vs-fragments
> A URL cannot have more than one fragment.